File: java/src/main/java/it/giannibombelli/wsc2026/payment/application/usecases/TransactionAccepting.java
package it.giannibombelli.wsc2026.payment.application.usecases;

import it.giannibombelli.wsc2026.common.utils.Require;

import it.giannibombelli.wsc2026.common.application.events.EventPublisher;
import it.giannibombelli.wsc2026.common.application.events.EventSubscriber;
import it.giannibombelli.wsc2026.payment.application.commands.AcceptTransaction;
import it.giannibombelli.wsc2026.payment.domain.events.PaymentEvent;
import it.giannibombelli.wsc2026.payment.domain.events.PaymentResultEvents;
import it.giannibombelli.wsc2026.payment.domain.events.TransactionAccepted;
import it.giannibombelli.wsc2026.payment.domain.payment.Payment;
import it.giannibombelli.wsc2026.payment.domain.policies.PaymentCompletion;
import it.giannibombelli.wsc2026.payment.domain.ports.PaymentRepository;

import static java.util.Objects.requireNonNull;

public final class TransactionAccepting implements EventSubscriber<TransactionAccepted> {
    private final PaymentRepository paymentRepository;
    private final EventPublisher<PaymentEvent> eventPublisher;
    private final PaymentCompletion paymentCompletion;

    public TransactionAccepting(PaymentRepository paymentRepository,
                                EventPublisher<PaymentEvent> eventPublisher,
                                PaymentCompletion paymentCompletion) {
        this.paymentRepository = requireNonNull(paymentRepository);
        this.eventPublisher = requireNonNull(eventPublisher);
        this.paymentCompletion = requireNonNull(paymentCompletion);
    }

    @Override
    public void on(TransactionAccepted event) {
        Require.requireArgument(event, "event");

        AcceptTransaction cmd = paymentCompletion.evaluate(event);
        Payment payment = paymentRepository.findById(cmd.aggregateId())
            .orElseThrow(() -> new IllegalArgumentException("payment not found"));

        PaymentResultEvents.PaymentAccepted accepted = payment.acceptTransaction(
            cmd.transactionId(),
            cmd.providerCompletedAt()
        );
        paymentRepository.save(payment);

        if (accepted != null) {
            eventPublisher.publish(accepted);
        }
    }
}

File: typescript/src/payment/application/usecases/transactionAccepting.ts
// Caso d'uso reattivo che accetta una transazione e completa il pagamento se possibile.

import type { EventPublisher } from '@/common/application/events/eventPublisher.js';
import type { EventSubscriber } from '@/common/application/events/eventSubscriber.js';
import { PaymentEvent } from '../../domain/events/paymentEvent.js';
import { TransactionAccepted } from '../../domain/events/transactionAccepted.js';
import { PaymentNotFoundException } from '../../domain/payment/paymentNotFoundException.js';
import { PaymentCompletion } from '../../domain/policies/paymentCompletion.js';
import type { PaymentRepository } from '../../domain/ports/paymentRepository.js';
import { requireArgument } from '@/common/utils/requireArgument.js';
import { requireDependency } from '@/common/utils/requireDependency.js';

export class TransactionAccepting implements EventSubscriber<TransactionAccepted> {
  private readonly _paymentRepository: PaymentRepository;
  private readonly _eventPublisher: EventPublisher<PaymentEvent>;
  private readonly _paymentCompletion: PaymentCompletion;

  constructor(
    paymentRepository: PaymentRepository,
    eventPublisher: EventPublisher<PaymentEvent>,
    paymentCompletion: PaymentCompletion
  ) {
    requireDependency(paymentRepository, "paymentRepository");
    requireDependency(eventPublisher, "eventPublisher");
    requireDependency(paymentCompletion, "paymentCompletion");
    this._paymentRepository = paymentRepository;
    this._eventPublisher = eventPublisher;
    this._paymentCompletion = paymentCompletion;
  }

  on(event: TransactionAccepted): void {
    requireArgument(event, 'event');

    const cmd = this._paymentCompletion.evaluate(event);
    const payment = this._paymentRepository.findById(cmd.aggregateId);
    if (payment === null) throw new PaymentNotFoundException();

    const accepted = payment.acceptTransaction(cmd.transactionId, cmd.providerCompletedAt);
    this._paymentRepository.save(payment);

    if (accepted !== null) {
      this._eventPublisher.publish(accepted);
    }
  }
}
